tests/components/broadlink/test_device.py

```diff
@@ -0,0 +1,92 @@
+"""Tests for Broadlink devices."""
+import broadlink.exceptions as blke
+import pytest
+
+from homeassistant.components.broadlink.const import DOMAIN
+from homeassistant.components.broadlink.device import BroadlinkDevice
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.setup import async_setup_component
+
+from . import pick_device
+
+from tests.async_mock import patch
+
+
+async def test_device_setup(hass):
+    """Test a successful setup."""
+    device = pick_device(1)
+    mock_api = device.get_mock_api()
+    mock_api.type = "RM4"
+    mock_entry = device.get_mock_entry()
+    device_entry = BroadlinkDevice(hass, mock_entry)
```

Comment:
Can we test this without accessing integration details in the tests?

Reply:
I think so, if I mock the update manager. I will see what I can do.

Reply:
I managed to do it with the following lines:
```python3
mock_update_manager = MagicMock()
mock_update_manager.coordinator.last_update_success = True
mock_update_manager.coordinator.async_refresh = AsyncMock()

with patch("broadlink.gendevice", return_value=mock_api), patch(
    "homeassistant.components.broadlink.device.get_update_manager",
    return_value=mock_update_manager
), patch.object(
    hass.config_entries, "async_forward_entry_setup"
) as mock_forward:
    assert await device_entry.async_setup() is True
```

But now we are not able to:
- test the update coordinator
- test whether we are forwarding the device to the correct platforms

I think the best approach would be to make the type a class member in the python-broadlink library and obtain it in the get_mock_api function.

Reply:
The goal should be to use the core to start the test and patch the library in the periphery, so we don't need to know details about the integration in the test.

We only need to check that the correct states are present in the state machine after setting up the integration with a device. We don't care about integration internal specifics.

Reply:
I moved the attribute to the mock. It's a bit cleaner now.
I can make this better in the future PR, but I would have to change the library.